contracts/finance/Vesting_2.sol

```diff
@@ -0,0 +1,271 @@
+        uint256 durationInPeriods;
+        uint256 cliffInPeriods;
+    }
+
+    struct Schedule {
+        BaseSchedule scheduleData;
+        uint256 exponent;
+    }
+
+    struct VestingData {
+        uint256 vestingStartTime;
+        address beneficiary;
+        address vestingToken;
+        uint256 vestingAmount;
+        uint256 paidAmount;
+        uint256 scheduleId;
+    }
+
+    uint256 public LINEAR_EXPONENT = 1;
+
+    uint256 internal _scheduleId;
+    uint256 internal _vestingId;
+
+    mapping(uint256 id => Schedule) internal _schedules;
+    mapping(uint256 id => VestingData) internal _vestings;
+    mapping(address beneficiary => EnumerableSet.UintSet) internal _beneficiaryIds;
+
+    // initialization
+    function __Vesting_init() internal onlyInitializing {
+        // for (uint256 i = 0; i < schedules_.length; i++) {
+        //     _createSchedule(schedules_[i]);
+        // }
+    }
+
+    function _createSchedule(Schedule memory schedule_) internal virtual returns (uint256) {
+        _validateSchedule(schedule_.scheduleData);
+
+        _schedules[++_scheduleId] = schedule_;
+
+        emit ScheduleCreated(_scheduleId);
+
+        return _scheduleId;
+    }
+
+    function _createSchedule(
+        BaseSchedule memory baseSchedule_
+    ) internal virtual returns (uint256) {
+        _validateSchedule(baseSchedule_);
+
+        _schedules[++_scheduleId] = Schedule({
+            scheduleData: baseSchedule_,
+            exponent: LINEAR_EXPONENT
+        });
+
+        return _scheduleId;
+    }
+
+    function _validateSchedule(BaseSchedule memory schedule_) internal pure virtual {
+        require(
+            schedule_.durationInPeriods > 0 && schedule_.secondsInPeriod > 0,
+            "VestingWallet: cannot create schedule with zero duration or zero seconds in period"
+        );
+    }
+
+    function getSchedule(uint256 scheduleId_) public view virtual returns (Schedule memory) {
+        return _schedules[scheduleId_];
+    }
+
+    function getVesting(uint256 vestingId_) public view virtual returns (VestingData memory) {
+        return _vestings[vestingId_];
+    }
+
+    // get vesting ids by beneficiary
+    function getVestings(address beneficiary_) public view virtual returns (VestingData[] memory) {
+        uint256[] memory _ids = _beneficiaryIds[beneficiary_].values();
+        VestingData[] memory _beneficiaryVestings = new VestingData[](_ids.length);
+
+        for (uint i = 0; i < _ids.length; i++) {
+            _beneficiaryVestings[i] = _vestings[i];
+        }
+
+        return _beneficiaryVestings;
+    }
+
+    // get vesting ids by beneficiary
+    function getVestingIds(address beneficiary_) public view virtual returns (uint256[] memory) {
+        return _beneficiaryIds[beneficiary_].values();
+    }
+
+    // get available amount to withdraw
+    function getWithdrawableAmount(
+        uint256 vestingId_,
+        uint256 timestampUpTo_,
+        uint256 timestampCurrent_
```

Comment:
Let's get rid of `timestampCurrent_` in every public and internal function. We may use it only in virtual ones for the trusted user to override

Reply:
Timestamps have to be handled by the SC itself. We may only use `timestampUpTo_` in `view` function for front end convenience